### PR TITLE
v2http: allow empty role for GET '/users'

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/5246.

### CODE REVIEW

**Comment by gyuho** [comment]
Previous:

```
./bin/etcdctl user add root:a
# auth: created user root

curl -L -X PUT  http://127.0.0.1:2379/v2/auth/users/root -d '{"user": "root", "grant": ["foo"]}'
# auth: updated user root

curl -L -u root http://127.0.0.1:2379/v2/auth/users
# {"message":"auth: Role foo does not exist."}
```


Now it returns `"role":""`:

```
./bin/etcdctl user add root:a
# User root created

curl -L -X PUT  http://127.0.0.1:2379/v2/auth/users/root -d '{"user": "root", "grant": ["foo"]}'
# auth: updated user root

curl -L -u root http://127.0.0.1:2379/v2/auth/users
# Enter host password for user 'root':
# {"users":[{"user":"root","roles":[{"role":"","permissions":{"kv":{"read":null,"write":null}}},{"role":"root","permissions":{"kv":{"read":["/*"],"write":["/*"]}}}]}]}

# server side
2016-05-04 13:58:53.028400 W | v2http: GetRole foo error (auth: Role foo does not exist.)
```

/cc @xiang90 @heyitsanthony 

Please review.

Thanks.

File: etcdserver/api/v2http/client_auth.go [hunk 1]
```diff
@@ -326,10 +326,10 @@ func (sh *authHandler) baseUsers(w http.ResponseWriter, r *http.Request) {
 		for _, roleName := range user.Roles {
 			var role auth.Role
 			role, err = sh.sec.GetRole(roleName)
-			if err != nil {
-				writeError(w, r, err)
-				return
+			if err != nil { // in case, role does not exist
+				plog.Warningf("GetRole %s error (%v)", roleName, err)
```
Comment (heyitsanthony) [inline etcdserver/api/v2http/client_auth.go]: should this `continue` instead of falling through and appending an empty role?

Comment (gyuho) [inline etcdserver/api/v2http/client_auth.go]: Right, will fix. Thanks.

Comment (gyuho) [inline etcdserver/api/v2http/client_auth.go]: Just fixed. PTAL. Thanks.

**Comment by heyitsanthony** [comment]
lgtm, deferring to @xiang90

File: etcdserver/api/v2http/client_auth.go (hunk 1 above)
Comment (xiang90) [inline etcdserver/api/v2http/client_auth.go]: i guess we should just remove the warning. this is nothing actually... etcd-users can assign a non-existing role to a user.

Comment (gyuho) [inline etcdserver/api/v2http/client_auth.go]: Sure, will remove.

File: e2e/v2_curl_test.go [hunk 2]
```diff
@@ -116,6 +116,29 @@ func TestV2CurlIssue5182(t *testing.T) {
 	}
 }
 
+func TestV2CurlIssue5246(t *testing.T) {
```
Comment (xiang90) [inline e2e/v2_curl_test.go]: this test does not really belong here. we should add a test in client_auth_test.go. e2e test is to ensure the correctness of etcdctl/etcd interaction, not to ensure the correctness of the server itself.

**Comment by gyuho** [comment]
@xiang90 I moved the test into `v2http` pkg. PTAL. Thanks.

**Comment by gyuho** [comment]
@xiang90 Now test only tests the `baseUser` handler (mock auth store change was needed as well). 

PTAL

File: etcdserver/api/v2http/client_auth_test.go [hunk 3]
```diff
@@ -361,6 +370,62 @@ func TestAuthFlow(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	defer resp.Body.Close()
+
+	var uc usersCollections
+	if err := json.NewDecoder(resp.Body).Decode(&uc); err != nil {
+		t.Fatal(err)
+	}
+	for _, u := range uc.Users {
+		if u.User == "root" {
```
Comment (xiang90) [inline etcdserver/api/v2http/client_auth_test.go]: first we need to ensure the user root does exist.

File: etcdserver/api/v2http/client_auth_test.go [hunk 4]
```diff
@@ -361,6 +370,62 @@ func TestAuthFlow(t *testing.T) {
 	}
 }
 
+func (sh *authHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	sh.baseUsers(w, r)
+}
+
+func TestIssue5246(t *testing.T) {
```
Comment (xiang90) [inline etcdserver/api/v2http/client_auth_test.go]: TestGetUserGrantedWithNonexistingRole

File: etcdserver/api/v2http/client_auth_test.go (hunk 3 above)
Comment (gyuho) [inline etcdserver/api/v2http/client_auth_test.go]: > first we need to ensure the user root does exist.

Can you explain more? Current output is:

> {"users":[{"user":"root","roles":[{"role":"root","permissions":{"kv":{"read":["/*"],"write":["/*"]}}}]}]}

I thought we want to return empty role just for the non-existing `foo`?

Comment (gyuho) [inline etcdserver/api/v2http/client_auth_test.go]: Ok nvm. You said `does exist`... :) Will fix.

Comment (xiang90) [inline etcdserver/api/v2http/client_auth_test.go]: this test does not ensure user root exist. it only fails if `root` exists and `foo` exists. it should also fail if `root` does not exist.

File: etcdserver/api/v2http/client_auth_test.go [hunk 5]
```diff
@@ -361,6 +370,62 @@ func TestAuthFlow(t *testing.T) {
 	}
 }
 
+func (sh *authHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Comment (xiang90) [inline etcdserver/api/v2http/client_auth_test.go]: i do not think we need this. use https://golang.org/pkg/net/http/#HandlerFunc

Comment (gyuho) [inline etcdserver/api/v2http/client_auth_test.go]: Without it, errors:

> ./client_auth_test.go:394: cannot use sh (type *authHandler) as type http.Handler in argument to httptest.NewServer:

https://golang.org/pkg/net/http/httptest/#NewServer

**Comment by gyuho** [comment]
All addressed. PTAL.

File: etcdserver/api/v2http/client_auth_test.go (hunk 5 above)
Comment (xiang90) [inline etcdserver/api/v2http/client_auth_test.go]: it should work.

```go
httptest.NewServer(http.HandlerFunc(sh.baseUsers))
```

Comment (gyuho) [inline etcdserver/api/v2http/client_auth_test.go]: Let me try use `HandlerFunc`

File: etcdserver/api/v2http/client_auth_test.go [hunk 6]
```diff
@@ -361,6 +370,67 @@ func TestAuthFlow(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	defer resp.Body.Close()
+
+	var uc usersCollections
+	if err := json.NewDecoder(resp.Body).Decode(&uc); err != nil {
+		t.Fatal(err)
+	}
+	rootExist := false
+	for _, u := range uc.Users {
```
Comment (xiang90) [inline etcdserver/api/v2http/client_auth_test.go]: just remove this range. get user root directly.

Comment (xiang90) [inline etcdserver/api/v2http/client_auth_test.go]: if len(us.Users) != 1 {t.fatal(...)}

u := us.Users[0]
if u.User != "root" {t.fatal()}
...

Comment (gyuho) [inline etcdserver/api/v2http/client_auth_test.go]: ok will fix.

**Comment by gyuho** [comment]
@xiang90 Actually, we can't check the user number because we use that mockStore for other tests, so AllUsers method returns 3 users now... https://github.com/coreos/etcd/blob/master/etcdserver/api/v2http/client_auth_test.go#L46

**Comment by xiang90** [comment]
> @xiang90 Actually, we can't check the user number because we use that mockStore for other tests, so AllUsers method returns 3 users now...

Can we fix it? I do not really care about that 3 actually. I should just get the user out of the users and do what we need to do instead of ranging over every users. That makes the code more readable.

**Comment by gyuho** [comment]
@xiang90 We can have another type of mock store for this test, or remove test cases using those 3 users. Which one do we prefer?

**Comment by xiang90** [comment]
@gyuho Which one is simpler. Or just keep the 3 there but get the root user out of it.

**Comment by gyuho** [comment]
@xiang90 I fixed the AllUsers for mock store to make it work for our new test. Works now.

PTAL.

Thanks!

**Comment by xiang90** [comment]
LGTM

**Comment by gyuho** [comment]
Thanks. Merging in greens.

**Pull request merged**